Feature/core 954 identity auth message

Tschuck opened the pull request.

Commit:
add identity checks to auth middleware

- [CORE-954]

Commit:
add `authMessage` identity address support

- [CORE-954]

Commit:
update core package versions

- [CORE-954]

Commit:
remove duplicated dependencies

- [CORE-954]

Commit:
Merge branch 'develop' into feature/CORE-954-identity-auth-message

Commit:
add account-identity / profile address cache to auth middleware

- [CORE-954]

Commit:
Merge branch 'feature/CORE-954-identity-auth-message' of github.com:evannetwork/edge-server-seed into feature/CORE-954-identity-auth-message

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
-    expect(isAuthenticated).to.be.eq(undefined)
-  })
+  runTestsWithConfig('Test signed requests with account based profile', testAccounts.oldAccount);
+  runTestsWithConfig('Test signed requests with identgity based profile', testAccounts.newAccount);
```
```suggestion
  runTestsWithConfig('Test signed requests with identity based profile', testAccounts.newAccount);
```

wulfraem replied:
```suggestion
  runTestsWithConfig('Test signed requests with identgity based profile', testAccounts.newAccount)
```

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
+        expect(isAuthenticated).to.be.eq(undefined)
+      })
+
+      it('reject when accessing unpermitted identity', async () => {
```
```suggestion
      it('should reject when accessing unpermitted identity', async () => {
```

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
+        connection.rawConnection.req = {
+          headers: await getAuthHeaders(testAccountRuntime, 0, '0x18aa46f4940817b132ade068f08d13df44e07220') // 6 minutes before
+        }
+        connection.rawConnection.req.he
```
```suggestion
```
???

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
+        expect(isAuthenticated).to.be.eq(undefined)
+      })
+
+      it('can retrieves error when auth time is to long ago', async () => {
```
```suggestion
      it('should return an error when auth time is too long ago', async () => {
```

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
+        expect(isAuthenticated).to.be.eq(true)
+      })
+
+      it('can retrieves error when auth failed', async () => {
```
```suggestion
      it('should return an error when auth failed', async () => {
```

S3bb1 commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
+        expect(api.id).to.be.ok()
+      })
+
+      it('can retrieve successfull auth status', async () => {
```
```suggestion
      it('should authenticate requests succesful', async () => {
```

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

Commit:
Update test/testSignedRequest.js

Co-Authored-By: Sebastian Dechant <sebastian.dechant@evan.team>

wulfraem commented on middlewares/authentication.js:
```diff
@@ -1,15 +1,23 @@
+  profileIndexAddress: '',
+  identity: { },
+  profile: { },
+};
```
```suggestion
}
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+  // if no identity was passed as auth header, check for old / new profile and try to load identity
+  const runtime = api.smartAgentCore.runtime;
+  if (!authComponents.EvanIdentity) {
+    const nullAddress = '0x0000000000000000000000000000000000000000';
```
```suggestion
    const nullAddress = '0x0000000000000000000000000000000000000000'
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+      const profileIndex = runtime.nameResolver.contractLoader
+        .loadContract('ProfileIndexInterface', authCache.profileIndexAddress)
+
+      // Check if a identity address has a underlying profile
```
```suggestion
      // Check if an identity address has an underlying profile
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+          { from: authComponents.EvanIdentity },
+        )
+
+      // if no profile could be found, its basically a old profile and we can reset the identity
```
```suggestion
      // if no profile could be found, it's basically an old profile and we can reset the identity
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -49,8 +116,8 @@ const authMiddleware = {
-  preProcessor: ({ connection }) => {
-    _ensureAuth(connection)
+  preProcessor: async ({ connection }) => {
+    await _ensureAuth(connection)
```
❓ doesn't this also work?
```suggestion
  preProcessor: _ensureAuth,
```

wulfraem commented on test/testSignedRequest.js:
```diff
@@ -29,112 +29,154 @@ const { createDefaultRuntime, Ipfs, utils: { getSmartAgentAuthHeaders } } =
-    expect(error).to.be.eq('Error: Signed message has been expired.')
-    expect(isAuthenticated).to.be.eq(undefined)
-  })
+  runTestsWithConfig('Test signed requests with account based profile', testAccounts.oldAccount);
```
```suggestion
  runTestsWithConfig('Test signed requests with account based profile', testAccounts.oldAccount)
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
   }
 
+  // if no identity was passed as auth header, check for old / new profile and try to load identity
+  const runtime = api.smartAgentCore.runtime;
```
```suggestion
  const runtime = api.smartAgentCore.runtime
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+      if (profileAddress === nullAddress) {
+        authComponents.EvanIdentity = authComponents.EvanAuth
+      } else {
+        authCache.profile[authComponents.EvanIdentity] = profileAddress;
```
```suggestion
        authCache.profile[authComponents.EvanIdentity] = profileAddress
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+  if (authComponents.EvanAuth !== authComponents.EvanIdentity) {
+    const verificationHolderContract = runtime.contractLoader.loadContract(
+      'VerificationHolder', authComponents.EvanIdentity,
+    );
```
```suggestion
    )
```

wulfraem commented on middlewares/authentication.js:
```diff
@@ -38,6 +46,65 @@ const _ensureAuth = (connection) => {
+      'keyHasPurpose',
+      runtime.nameResolver.soliditySha3(authComponents.EvanAuth),
+      '1',
+    );
```
```suggestion
    )
```

wulfraem commented on initializers/smart-agent-core.js:
```diff
@@ -36,6 +37,13 @@ module.exports = class SmartAgentCore extends Initializer {
+
+    // objects and values used outside initializer
+    api.smartAgentCore = smartAgentCore
+
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on config/smart-agent-core.js:
```diff
@@ -1,6 +1,11 @@
+  ethAccounts: {
+    '0x66D487419Ac030993Dba3D893Fc99bdafDCf60bB': 'cf316cf854f6d4ae7bd7c28ad124ae6ff95e3f4f7439e67a85689c0b8f632c13'
+  },
+  smartAgentCore: () => {
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on config/smart-agent-core.js:
```diff
@@ -1,6 +1,11 @@
+  },
+  smartAgentCore: () => {
     return {
+      ethAccount: '0x66D487419Ac030993Dba3D893Fc99bdafDCf60bB',
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on config/smart-agent-core.js:
```diff
@@ -18,8 +23,13 @@ exports['default'] = {
+  ethAccounts: {
+    '0x992b98e59CFcC47d982F0eB8Bbb2B00D85EDCce2': '874146ca958a4937c62e98b4043d4c392438285a9e06f80798dc52bf0c0e9804'
+  },
+  smartAgentCore: () => {
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on config/smart-agent-core.js:
```diff
@@ -18,8 +23,13 @@ exports['default'] = {
+  },
+  smartAgentCore: () => {
     return {
+      ethAccount: '0x992b98e59CFcC47d982F0eB8Bbb2B00D85EDCce2',
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on initializers/smart-agent-core.js:
```diff
@@ -16,6 +16,7 @@
 
 'use strict'
 const { Initializer, api } = require('actionhero')
+const { createDefaultRuntime } = require('@evan.network/api-blockchain-core')
```
(belongs to question group 1)
❓ Does the seed need its own runtime? Derived smart agents will have a running runtime and the seed itself will most probably never use its own `runtime` and auth. So every Serve would have an extra runtime with default accounts and a key exchange listener, that does nothing or am I wrong?
imho we could move the `_ensureAuth` into the class scope and use the `runtime` from the current instance... or? ^^
```suggestion
```

wulfraem commented on VERSIONS.md:
```diff
@@ -2,6 +2,7 @@
 
 ## Next Version
 ### Features
+- add `authMessage` identity address support
```
```suggestion
- add `EvanIdentity` authentication header to support authentication via identities
```

wulfraem reviewed (changes requested):
Review comments, that start with "(belongs to question group 1)" belong to the same group.
Merging these also need further updates to the code. ^^

Commit:
Update middlewares/authentication.js

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>

Commit:
Update middlewares/authentication.js

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>

Commit:
use correct identity auth middleware

- remove smart-agent-core account config
- add `ensureIdentityAuth` to `middlwares/authentication`
- add `registerAuthMiddleware` to `SmartAgent` class for enabling identity authentication with custom auth middlewares
- enhance `readme.md`
- [CORE-954]

Commit:
merge

Commit:
remove async from default auth middleware

- [CORE-954]

Commit:
adjust versions

- [CORE-954]

wulfraem commented on README.md:
```diff
@@ -37,3 +37,102 @@ SmartAgents work more or less the same as the SmartAgents in vanilla [ActionHero
 different puposes, But usually it is just a SHA3 hash of an accountID or of a descriptive string.
+
+## Register SmartAgent
+For registering a SmartAgent, create a initializer and add something like the following:
```
```suggestion
For registering a SmartAgent, create an initializer and add something like in the example below. This creates a new instance of your SmartAgent and makes it available for other actionhero components at the `api` (in this case at `api.smartAgentTest`. As this SmartAgent extends `api.smartAgents.SmartAgent`, it will have a property called `runtime`, that can be used for interaction with evan.network (see [API documentation](https://api-blockchain-core.readthedocs.io/en/latest/index.html) for details).
```

wulfraem commented on README.md:
````diff
@@ -37,3 +37,102 @@ SmartAgents work more or less the same as the SmartAgents in vanilla [ActionHero
+module.exports = Authenticated
+```
+
+The latest updates of the `@evan.network/api-blockchain-core` also provides the possibility to check if the passed `EvanAuth` address is allowed to interact on behalf of the passed `EvanIdentity`. To enable this check, you need to register your own authentication middleware from your smart agent instance and pass a valid `@evan.network/api-blockchain-core` runtime.
````
```suggestion
The latest updates of the `@evan.network/api-blockchain-core` also provide the possibility to check if the passed `EvanAuth` address is allowed to interact on behalf of the passed `EvanIdentity`. To enable this check, you need to register your own authentication middleware from your smart agent instance and pass a valid `@evan.network/api-blockchain-core` runtime.
```

wulfraem commented on lib/smart-agent.js:
```diff
@@ -181,6 +182,25 @@ module.exports = class SmartAgent {
 
+  /**
+   * Registers an auth middleware with the current runtime with the smart agent config name (e.g.
+   * ensureSearchAuth) Usual `ensureEvanAuth` middleware has no runtime context and cannot ensure
```
```suggestion
   * ensureSearchAuth). Usually `ensureEvanAuth` middleware has no runtime context and cannot ensure
```

wulfraem commented on lib/smart-agent.js:
```diff
@@ -181,6 +182,25 @@ module.exports = class SmartAgent {
+  /**
+   * Registers an auth middleware with the current runtime with the smart agent config name (e.g.
+   * ensureSearchAuth) Usual `ensureEvanAuth` middleware has no runtime context and cannot ensure
+   * identity authorization.
```
```suggestion
   * identity authentication.
```

Commit:
use full authName for registerAuthMiddleware

- [CORE-954]

wulfraem reviewed (changes requested):
Also as discussed, update middleware naming behavior. ^^

Commit:
Update README.md

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>

Commit:
Update README.md

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>

Commit:
Update lib/smart-agent.js

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>

Commit:
Update lib/smart-agent.js

Co-Authored-By: wulfraem <wulfraem@users.noreply.github.com>
